controller/scheduler/scheduler.go

```diff
@@ -0,0 +1,1216 @@
+	logger.Info("triggering rectify", "key", key)
+	s.rectifyBatch[key] = struct{}{}
+	select {
+	case s.rectify <- struct{}{}:
+	default:
+	}
+}
+
+func (s *Scheduler) updateFormation(f *ct.Formation) (*Formation, error) {
+	ef, err := utils.ExpandFormation(s, f)
+	if err != nil {
+		return nil, err
+	}
+	return s.changeFormation(ef)
+}
+
+func (s *Scheduler) startJob(req *JobRequest) (err error) {
+	log := fnLogger("job.type", req.Type)
+	log.Info("starting job", "job.restarts", req.restarts, "request.attempts", req.attempts)
+	s.jobs.SetState(req.JobID, JobStateStopped)
+	// Copy on Write
+	newReq := req.Clone()
+	newReq.HostID = ""
+	newReq.JobID = random.UUID()
+	newReq.state = JobStateRequesting
+	defer func() {
+		if err != nil {
+			if newReq.attempts >= maxJobAttempts {
+				log.Error("error starting job, max job attempts reached", "err", err)
+			} else {
+				log.Error("error starting job, trying again", "err", err)
+				newReq.attempts++
+				s.jobs[newReq.JobID] = newReq.Job
+				time.AfterFunc(jobAttemptInterval, func() {
+					s.jobRequests <- newReq
+				})
+			}
+		} else {
+			s.jobs[newReq.JobID] = newReq.Job
+		}
+	}()
+
+	log.Info("determining best host for job")
+	host, err := s.findBestHost(newReq.Formation, newReq.Type)
+	if err != nil {
+		log.Error("error determining best host for job", "err", err)
+		return err
+	}
+	hostID := host.ID()
+	newReq.HostID = hostID
+
+	config := jobConfig(newReq, hostID)
+	newReq.JobID = config.ID
+
+	// Provision a data volume on the host if needed.
+	if newReq.needsVolume() {
+		log.Info("provisioning volume")
+		if err := utils.ProvisionVolume(host, config); err != nil {
+			log.Error("error provisioning volume", "err", err)
+			return err
+		}
+	}
+
+	log.Info("requesting host to add job", "host.id", hostID, "job.id", config.ID)
+	if err := host.AddJob(config); err != nil {
+		log.Error("error requesting host to add job", "err", err)
+		return err
+	}
+	return nil
+}
+
+func (s *Scheduler) stopJob(req *JobRequest) (err error) {
+	log := fnLogger("req.host.id", req.HostID, "req.job.id", req.JobID, "job.type", req.Type)
+	log.Info("stopping job")
+	defer func() {
+		if err != nil {
+			log.Error("error stopping job", "err", err)
+		}
+	}()
+
+	var job *Job
+	if req.JobID == "" {
+		formationKey := utils.FormationKey{AppID: req.AppID, ReleaseID: req.ReleaseID}
+		typJobs := s.jobs.GetStoppableJobs(formationKey, req.Type)
+		if len(typJobs) == 0 {
+			e := fmt.Sprintf("no %s jobs running", req.Type)
+			log.Error(e)
+			return errors.New(e)
+		}
+		job = typJobs[0]
+		for _, j := range typJobs {
+			if j.startedAt.After(job.startedAt) {
+				job = j
+			}
+		}
+		log.Info("selected job for termination", "job.id", job.JobID, "job.host.id", job.HostID)
+	} else {
+		var ok bool
+		job, ok = s.jobs[req.JobID]
+		if !ok {
+			e := "unknown job"
+			log.Error(e)
+			return errors.New(e)
+		}
+	}
+	s.jobs.SetState(job.JobID, JobStateStopped)
```

Comment:
Setting the job's state to `JobStateStopped` here means that when we get a job event indicating the job is down, we don't persist it to the controller because the job is already in the stopped state, so it seems only `SyncJobs` will cause stopped events to be persisted.